### ISSUE TITLE
Docs for building midlets using FOSS tools and SDKs

### ISSUE THREAD

**Issue by KOLANICH** [j2me-preservation/j2me-preservation#2]
Could anyone, please, write an article about state of the art way of building midlets using as much free SDKs and tools (i.e., can I use OpenJDK from distro packages, also, can other JVM-based langs be used?) as possible in *nix-based OSes ?

@nikita36078 @vadosnaprimer @szatkus @hex007 @XerTheSquirrel @jbanes @ichisadashioko @artem-frolov

**Comment by minexew** [comment]
FWIW, I'm trying to compile Sun's phoneME VM from the archived SVN repository, and I'll definitely write it up if I succeed.

As for your question, I *think* OpenJDK 8 still supports `-source 1.3 -target 1.1`, which would be compliant with CLDC, but you will also need the preverifier, and of course the MIDP classpath.

**Comment by ichisadashioko** [comment]
- I think we can create a dummy implementation of all the APIs by putting `throw new UnsupportedOperation()` inside those methods.
- Use an old openjdk release (1.4) to compile "our source code" with `javac` into `.class` files.
- Create a zip archive with "our class files" excluding the midlet classes + the manifest file
- I think the archive can be use in every emulators now.

**Comment by ichisadashioko** [comment]
The openjdk 15 for Windows is just an tarball. We can just extract the tarball and set appropriate environment variables. I'm not sure if that is the case with openjdk for Linux and older releases of openjdk.

I tried to install Ubuntu 10.x, 8.x the other day because I think those were the supported distro during midlets development. However, that didn't go smoothly as I expected.

- relocated repositories URLs
- expired SSL certificate -> cannot use `git` to pull anything and access the internet (can only access http localhost from my other machines).
- outdated libraries so that we cannot compile most new version of most softwares from source.

**Comment by vadosnaprimer** [comment]
> FWIW, I'm trying to compile Sun's phoneME VM from the archived SVN repository, and I'll definitely write it up if I succeed.

Are you using the archived docs too?
https://phonej2me.github.io/content/phoneme_platforms.html#phonemefeature

**Comment by minexew** [comment]
> > FWIW, I'm trying to compile Sun's phoneME VM from the archived SVN repository, and I'll definitely write it up if I succeed.
> 
> Are you using the archived docs too?
> https://phonej2me.github.io/content/phoneme_platforms.html#phonemefeature

Not really, since they only speak about Win32, but I'm trying to build on Linux (Debian 5 i386, gcc 4.3). Also, I haven't found the sources for phoneME Feature MR3/MR4, only MR2.

This is the source tree: https://github.com/j2me-preservation/phoneME-GP2X-SDL

**Comment by XerTheSquirrel** [comment]
> Could anyone, please, write an article about state of the art way of building midlets using as much free SDKs and tools (i.e., can I use OpenJDK from distro packages, also, can other JVM-based langs be used?) as possible in *nix-based OSes ?

You could technically use SquirrelJME, _if you were to finish and allow for publishing to Maven Central or your local Maven repository which I rather set to the side for now_, to build whatever program. Although SquirrelJME is CLDC 8, MIDP 3, and MEEP 8 you would need to avoid using any APIs that it uses.

From experience the best target to use when compiling would be to target Java 7, from that point the classes would have to be processed due to some difference in the class format. But generally the following restrictions and bonuses apply:

 * You can use generics, because they are made to be backwards compatible and effectively are meaningless to the JVM.
 * You cannot use `invokedynamic`.
 * With Java 7 you are getting all of the restrictions that were inherited from Java ME, such as there being no "sub-routines" (`jsr` and `ret` instructions).

One thing you will need to do is convert Java 7's `StackMapTable` to the old `StackMap` that Java ME used. So the `StackMapTable` that currently exists has been derived from Java ME's `StackMap` which includes all of its restrictions and otherwise such as every instruction having a unique type state. So what the difference between these two is just the newer `StackMapTable`, which is the same in Java ME 8 and Java 7, is that it is a more compact and an adjusted format to be this way. The older legacy `StackMap` format contains the same information except for evrey possible instruction it will have information on the entire state of local and stack types.

Eventually at some point, I am thinking of making a Gradle plugin which would be capable of converting JARs and classes to the older J2ME with the assistance of SquirrelJME's libraries.

**Comment by XerTheSquirrel** [comment]
As an update I uploaded the SquirrelJME modules to the OSS SonaType Snapshot Nexus Repository. The base for the modules are `cc.squirreljme.modules:<module>:0.3.0-SNAPSHOT`.

![image](https://user-images.githubusercontent.com/11927855/103164875-c4808180-47de-11eb-859a-fa75e120f322.png)
